### PR TITLE
Separate file acquisition from file parsing

### PR DESCRIPTION
**Pull request by iwahbe**
For a given resource, the bridge:
1. Finds the file associated with the resource.
2. Parses the file into `entityDocs`

Currently, steps (1) and (2) are tied together in one function: `getDocsForResource`. This PR encapsulates (1) behind an interface (`DocsSource`) which is then passed to `getDocsForResource`. This makes testing much easier, and also simplifies our code.

---

The only user observable change in functionality is a much more obvious and useful error message when unable to find the upstream repository. The new warning looks like this:

```
warning: Unable to find the upstream provider's documentation:

The upstream repository is expected to be at "github.com/vancluever/terraform-provider-acme".

If the expected path is not correct, you should check the values of these fields (current values shown):

tfbridge.ProviderInfo{
        GitHubHost:              "github.com",
        GitHubOrg:               "vancluever",
        Name:                    "acme",
        TFProviderModuleVersion: "",
}

The original error is: error running 'go mod download -json github.com/vancluever/terraform-provider-acme' in "/Users/ianwahbe/go/src/github.com/pulumiverse/pulumi-acme/provider" dir for module: exit status 1

Output: {
        "Path": "github.com/vancluever/terraform-provider-acme",
        "Error": "module github.com/vancluever/terraform-provider-acme: not a known dependency"
}
```

Fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1365

### CODE REVIEW

File: pf/tests/internal/testprovider/tls.go [hunk 1]
```diff
@@ -54,13 +54,14 @@ func TLSProvider() tfbridge.ProviderInfo {
...
-		Homepage:    "https://pulumi.io",
-		Repository:  "https://github.com/pulumi/pulumi-tls",
+		Name:             "tls",
+		P:                tfpf.ShimProvider(tlsshim.NewProvider()),
+		Description:      "A Pulumi package to create TLS resources in Pulumi programs.",
+		Keywords:         []string{"pulumi", "tls"},
+		License:          "Apache-2.0",
+		Homepage:         "https://pulumi.io",
+		Repository:       "https://github.com/pulumi/pulumi-tls",
+		UpstreamRepoPath: ".",
```
Comment (mjeffryes) [inline pf/tests/internal/testprovider/tls.go:64]: Is this a new required field that all bridged providers will need to add?

Comment (iwahbe) [reply pf/tests/internal/testprovider/tls.go:64]: No. It is an existing field, and it is not required. I wanted some of our test providers to have their docs lookup path in valid go repos, while unaltered providers have their docs lookup path in invalid go repos.

Existing providers will not need to take any action after this PR merges.

File: pkg/tfgen/generate_schema_test.go [hunk 2]
```diff
@@ -41,24 +45,34 @@ import (
...
-	bridgetesting.AssertEqualsJSONFile(t, "test_data/regress-minirandom-schema.json", schema)
+	bridgetesting.AssertEqualsJSONFile(t, expectedFile, schema)
 }
 
 func TestNestedMaxItemsOne(t *testing.T) {
 	provider := testprovider.ProviderMiniCloudflare()
+
+	expectedFile := abs(t, "test_data/regress-minicloudflare-schema.json")
+	enterProviderModule(t,
+		"github.com/terraform-providers/terraform-provider-cloudflare v1.0.0")
```
Comment (mjeffryes) [inline pkg/tfgen/generate_schema_test.go:74]: Why do these tests now have to change their working dir to appear as if they are in the provider directory?

File: pkg/tfgen/source.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+func (gh *gitRepoSource) getResource(rawname string, info tfbridge.ResourceOrDataSourceInfo) (*DocFile, error) {
+	return gh.getFile(rawname, info, ResourceDocs)
+}
+
+func (gh *gitRepoSource) getDatasource(rawname string, info tfbridge.ResourceOrDataSourceInfo) (*DocFile, error) {
+	return gh.getFile(rawname, info, DataSourceDocs)
+}
+
+func (gh *gitRepoSource) getFile(
```
[the remark this answers is not in the document]
Comment (mjeffryes) [reply pkg/tfgen/source.go:92]: Since you've gone to the trouble of extracting this to it's own interface, it would be nice to add some units for it.

**Review by mjeffryes — COMMENTED** [review]
It feels like we're missing a test that clearly shows the new behavior (ie. x used to err, but now it just prints a warning)?

I know this is mostly just moving functions around to different locations, but there's enough small changes that I don't feel qualified to give the give the approving review given that I'm not super familiar with the context yet.

**Comment by t0yv0** [comment]
Can we please have refactors and fixes/behavior changes as separate PRs and try to keep them smaller? Thank you.

File: pkg/tfgen/source.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+)
+
+type DocsSource interface {
+	getResource(rawname string, info tfbridge.ResourceOrDataSourceInfo) (*DocFile, error)
```
Comment (t0yv0) [inline pkg/tfgen/source.go:29]: Is it not strange that a DocSource needs to know `info`? What is it reading from it? I see it's reading `info.GetDocs()` - is it a good idea for DocsSource's job to be merging overrides from info? Consider if you had Git-based and some other say AI-based DocsSource, both would have to implement overrides.

**Comment by iwahbe** [comment]
Closing in favor of https://github.com/pulumi/pulumi-terraform-bridge/pull/1387 and https://github.com/pulumi/pulumi-terraform-bridge/pull/1388.

**Pull request closed without merging**